### PR TITLE
Add support for symbolic nodes and box limits

### LINKED ISSUE #13323: Layout is broken when using `RangeTool` and other weird behavior
This is a collection of bugs, related partially to issue #13315:

* top plot's layout doesn't align correctly when ranges are invalid
* bottom plot's layout doesn't align correctly after reset
* second reset fixes the layout
* the overlay can appear outside of plot's frame (on and off)
* the overlay's size changes at some point when partially off-frame

https://github.com/bokeh/bokeh/assets/27475/de83b2bc-92b6-432e-88cc-e5ef545b23a1

### PR DESCRIPTION
**Pull request by mattpap**
The goals of this PR are:

1. Introduce a rudimentary symbolic coordinate system that would allow users to reference certain well known but hard to access coordinates like e.g. frame corners. This is required to generalize `null` case in `BoxAnnotation.{left,right,top,bottom}`.
2. Implement movement/resize limits on `BoxAnnotation`, which would fix the regression in `RangeTool` introduced in PR #12468.

TODO:

- [x] handle inverted boxes
- [x] add tests

https://github.com/bokeh/bokeh/assets/27475/a907e001-73f3-4e9d-9332-fcf38aaa55a9

fixes #13323

### CODE REVIEW

**Comment by mattpap** [comment]
Tests are still needed, but is tentatively ready for review. The main feature (regression fix really) is support for movement and resize limits on `BoxAnnotation` and thus `RangeTool`'s overlay. Symbolic node support is really primitive and just the bare minimum to allow for a sensible configuration API for this particular purpose. This feature is expected to be experimental for a while (`Node` model is marked accordingly).

**Review by ianthomas23 — APPROVED** [review]
I've only given the code a quick scan, but it looks good so far.

File: src/bokeh/core/property/bases.py [hunk 1]
```diff
@@ -247,19 +248,19 @@ def matches(self, new: T, old: T) -> bool:
...
         """
         import numpy as np
-        import pandas as pd
 
         if isinstance(new, np.ndarray) or isinstance(old, np.ndarray):
             return np.array_equal(new, old)
 
-        if isinstance(new, pd.Series) or isinstance(old, pd.Series):
-            return np.array_equal(new, old)
+        if uses_pandas(new) or uses_pandas(new):
```
Comment (bryevdv) [inline src/bokeh/core/property/bases.py]: Was this supposed to be
```suggestion
        if uses_pandas(new) or uses_pandas(old):
```
?

File: src/bokeh/models/annotations/geometry.py [hunk 2]
```diff
@@ -62,6 +68,10 @@
     "Whisker",
 )
 
+Coordinate = Either(CoordinateLike, Instance(Node))
+
+NaN = float("nan")
```
Comment (bryevdv) [inline src/bokeh/models/annotations/geometry.py]: This seems like it should go in a util module (and replace many duplicate instances of the same) but and issue will suffice if you don't want to add that change to this PR

File: src/bokeh/models/coordinates.py [hunk 3]
```diff
@@ -68,10 +78,57 @@ def __init__(self, *args, **kwargs) -> None:
...
+
+    This can be either a concrete model that can provide its coordinates (e.g.
+    a renderer, a frame or a canvas) or an implicit target defined by the
+    enum, which is resolved as the nearest parent of the given type. If the
+    provider cannot be determined or it isn't able to provide coordinates,
+    then the node resolved to an invalid coordinate (with x and y components
+    being ``NaN``).
+    """)
+
+    term = Required(String(), help="""
```
Comment (bryevdv) [inline src/bokeh/models/coordinates.py]:
```suggestion
    term = Required(String, help="""
```
would be more consistent with existing usage elsewhere

Comment (bryevdv) [reply src/bokeh/models/coordinates.py]: Also "term" doesn't really say anything to me, at all. If anything it suggests something more like a time duration ("semester term") What about just "symbol" instead?

Comment (mattpap) [reply src/bokeh/models/coordinates.py]: We have, for example, `:term:` directive in documentation. I took this name from logic programming. However, "symbol" does indeed seem more meaningful in this context.

File: src/bokeh/models/annotations/geometry.py (hunk 2 above)
Comment (mattpap) [reply src/bokeh/models/annotations/geometry.py]: Apparently there is now `math.nan` and `math.inf`, so I went ahead and used that and normalized all similar usage.

Comment (mattpap) [reply src/bokeh/models/annotations/geometry.py]: I moved `Coordinate` to `bokeh.models.common.properties`. Can't use `bokeh.model.util` due to circular imports.

**Comment by mattpap** [comment]
This is now fully completed.

**Pull request merged**